pep-0613.rst

```diff
@@ -0,0 +1,173 @@
+Status: Draft
+Type: Informational
+Content-Type: text/x-rst
+Created: 21-Jan-2020
+Post-History: 21-Jan-2020
+
+
+Abstract
+========
+
+Type aliases are user-specified types which may be as complex as any type hint, and are specified with a simple variable assignment on a module top level.
+
+This PEP formalizes a way to explicitly declare an assignment as a type alias.
+
+Motivation
+==========
+
+Type aliases are declared as top level variable assignments. In `PEP 484 <https://www.python.org/dev/peps/pep-0484/#type-aliases/>`_, the distinction between a valid type alias and a global variable was implicitly determined: if a top level assignment is unannotated, and the assigned value is a valid type, then the name being assigned to is a valid type alias. Otherwise, that name is simply a global value that cannot be used as a type hint.
+
+These implicit type alias declaration rules create confusion when type aliases involve forward aliases, invalid types, or violate other restrictions enforced on type alias declaration. Because the distinction between an unannotated value and a type alias is implicit, ambiguous or incorrect type alias declarations implicitly default to a valid value assignment. This creates expressions that are impossible to express as type aliases and punts error diagnosis of malformed type aliases downstream.
+
+The following examples each include an illustration of some of the suboptimal or confusing behaviors resulting from existing implicit alias declarations.
+We also introduce explicit aliases of the format :code:`TypeName: TypeAlias = Expression` here for the sake of comparison, but the syntax is discussed in further detail in later sections.
+
+Forward References:
+*******************
+
+::
+
+  MyType = "ClassName"
+  def foo() -> MyType: ...
+
+This code snippet should not error so long as :code:`ClassName` is defined later on. However, a type checker is forced to throw spurious errors that (1) :code:`MyType` is an unannotated global string, and (2) :code:`MyType` is not a valid type.
+
+::
+
+  MyType: TypeAlias = “ClassName”
+  def foo() -> MyType: ...
+
+Explicit aliases remove ambiguity so neither of the above errors will be thrown. Additionally, if something is wrong with :code:`ClassName` (i.e., it’s not actually defined later), the type checker can throw an error.
+
+
+Error Messaging:
+****************
+
+::
+
+  MyType = InvalidType
+  MyType = MyGeneric(int)  # i.e., intention was MyGeneric[int]
+
+A type checker should warn on this code snippet that :code:`InvalidType` is not a valid type, and therefore cannot be used to annotate an expression or to construct a type alias. Instead, type checkers are forced to throw spurious errors that (1) :code:`MyType` is a global expression missing an annotation, and (2) :code:`MyType` is not a valid type in all usages of :code:`MyType` across the codebase.
+
+::
+
+  MyType: TypeAlias = InvalidType
+  MyType: TypeAlias = MyGeneric(int)
```

Comment:
I'm confused. These look like two different examples somehow merged into one -- I assume you can't declare the same name as a `TypeAlias` twice?

Would this section work if the type aliases were named `MyType1` and `MyType2`?

(It's especially confusing since in the previous example, the two lines of the example had to occur together to illustrate the problem, while here there seem to be two independent lines that each illustrate a different variant of the problem.)

Reply:
Ah yes, I'll rename the variables so it's clear they're two separate examples in one code snippet.